### PR TITLE
Spacer : Added color support

### PR DESCRIPTION
**Pull request by benazeer-ben**
## What?

Adding color support for Spacer block.
Part of https://github.com/WordPress/gutenberg/issues/43245

## Why?

Spacer block is missing color support. 

## How?

Adds the color block support with bg color settings via block.json

## Testing Instructions

- Navigate to template/page edit.
- Then add spacer block and apply the background color .
- Verify that background color display correctly in both the editor and frontend.

## Screenshots or screencast 

Backend:
spacercolorbackend

Frontend:
spacercolorfrontend

### CODE REVIEW

**Comment by ramonjd** [comment]
Thanks for the PR! 

Here I'm wondering if adding visual styles to the spacer block is contradictory to the block's purpose:

> [Add white space between blocks and customize its height.](https://github.com/WordPress/gutenberg/blob/f6e9fccc236e11af1f114509ce92c84cf0dbacc1/packages/block-library/src/spacer/block.json#L6-L7)

If the block can have a background color then it ventures beyond its one-dimensional job of creating just "space".

Rather, there are alternative container blocks that can the same thing, e.g., Group block:


```html
<!-- wp:group {"style":{"dimensions":{"minHeight":"249px"}},"backgroundColor":"luminous-vivid-orange","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-luminous-vivid-orange-background-color has-background" style="min-height:249px"></div>
<!-- /wp:group -->
```

For that reason I'd probably skip adding any such block supports to the space block.

That's just my 2c.

**Review by aaronrobertshaw — COMMENTED** [review]
Appreciate the great PR descriptions for all these block support PRs @benazeer-ben, thank you 🙇 

This PR is testing as advertised for me. I didn't run into any issues.

✅ Global Styles apply correctly in the editor and on the frontend.
✅ Block support styles override the Global Styles as expected on both the frontend and editor

My understanding is that the spacer block was supposed to just be for layout and spacing purposes. I'm not sure it makes sense to support colors. 

I don't hold that opinion strongly so if others feel it's valuable, I'd be happy to change my mind. 

I've added the `Needs Design Feedback` label to this one to get a design perspective.

_Edit: Missed Ramon's reviewed this already 🤦_ 

> Here I'm wondering if adding visual styles to the spacer block is contradictory to the block's purpose:

That echoes my thinking just way more eloquently put 🙂

**Review by t-hamano — COMMENTED** [review]
Personally, I've never wanted to apply color to the Spacer block 😅

It would be nice to have a concrete example of why color support is needed.

**Comment by karthick-murugan** [comment]
IMO, Adding color to the Spacer block helps users visually identify spacing elements within the editor. This is particularly useful for complex layouts where empty space can be difficult to discern without a visible indication.

**Comment by aaronrobertshaw** [comment]
> IMO, Adding color to the Spacer block helps users visually identify spacing elements within the editor. This is particularly useful for complex layouts where empty space can be difficult to discern without a visible indication.

The adoption of color supports means that these "spacing elements" cease to be just that. They become visual elements as the colors get applied. Those colors are then visible not just to the user editing layouts but also the site's visitors on the frontend.

There are a few tools at the user's disposal to identify spacing elements e.g. on hover outlines, list view sidebar etc. 

The proposed use case might be better addressed through its own dedicated issue. The color block supports aren't a suitable solution to that problem.

My vote would be to close this PR.

**Comment by hanneslsm** [comment]
Being late to the party, I'd like to add that I am in favor of adding design tools also to the spacer block, because it dramatically enhances the experiences of how decorative elements can be build. I stated that in https://github.com/WordPress/gutenberg/issues/43247#issuecomment-2434951175 already.

With the official call for weird themes, I think we should consider reopening this PR
https://wordpress.org/news/2025/01/wordpress-themes-need-more-weird-a-call-for-creative-digital-homes/

**Comment by hanneslsm** [comment]
> Rather, there are alternative container blocks that can the same thing, e.g., Group block:

This argument is not valid by the way, because the group block is not rendered on the front when it's empty. The spacer block is.

**Comment by hanneslsm** [comment]
Related discussion: https://github.com/WordPress/gutenberg/pull/66443

**Comment by t-hamano** [comment]
> This argument is not valid by the way, because the group block is not rendered on the front when it's empty. The spacer block is.

Even if the Group block is empty, it should still be rendered on the frontend, like this:

```html
<div class="wp-block-group has-global-padding is-layout-constrained wp-block-group-is-layout-constrained"></div>
```

That is, you can apply border, padding, min-height, and colors to the Group block to make it look more decorative.

**Comment by hanneslsm** [comment]
> Even if the Group block is empty, it should still be rendered on the frontend, like this:

My bad! Thanks for pointing it out - I missed adding the padding. 
Still, I think it'd be a better experience when it'd be possible to change the style of the spacer.

**Comment by hanneslsm** [comment]
> Still, I think it'd be a better experience when it'd be possible to change the style of the spacer.

One argument for that is that the site editor displays the spacing "wrong" when it comes to an empty group with a padding.

Example:
A group with "x-large" padding and a Spacer with "x-large" height
A group with "small" padding and a Spacer with "small" height

CleanShot 2025-02-19 at 14 54 58@2x

On the font-end it's rendered correctly.
CleanShot 2025-02-19 at 14 55 56@2x


There was so much effort put into making the editor like the front end but still we fail on the simple tasks.
Not to mention that when selecting the parent group block the height of the inner groups with the padding changes again, which totally adds to confusing and bad UX.

https://github.com/user-attachments/assets/70a259fb-ff62-4c76-baf2-40d064770024

**Comment by hanneslsm** [comment]
And one more (sorry) but the argument 
> My understanding is that the spacer block was supposed to just be for layout and spacing purposes. I'm not sure it makes sense to support colors.

is in my opinion also invalid because you could say the same about the group: The group block was supposed to group content. If you only need a visual spacer (no matter which color) you should use the spacer block.

**Comment by benazeer-ben** [comment]
Based on this new [ticket](https://github.com/WordPress/gutenberg/issues/69263), reopening this PR.

More suggestions are welcome to help finalize this PR.

**Comment by aaronrobertshaw** [comment]
> Not to mention that when selecting the parent group block the height of the inner groups with the padding changes again, which totally adds to confusing and bad UX

I could replicate this and it appears related to a style adding margins to elements with the `.has-background` class and an inner `block-list-appender`. This is something that can be improved regardless of the proposal to add block supports to the Spacer block.

```
[data-type="core/group"].is-selected .has-background .block-list-appender{
  margin-bottom:18px;
  margin-top:18px;
}
```

> is in my opinion also invalid because you could say the same about the group: The group block was supposed to group content. 

I appreciate you challenging the current thinking. 

Here, though, the Group block isn't just spacing and layout, it's thought of more as a container or box. In the same way, a physical box can be decorated, so too could this container block. 

A spacer block is intended to be in effect negative space, the absence of something. 

> If you only need a visual spacer (no matter which color) you should use the spacer block.

I don't think this is correct.

In most cases, the visual space is a matter of presentation and shouldn't be achieved through the addition of another DOM element (which is what happens with the Spacer block). It would be better to leverage block spacing supports or padding/margin support, depending on the situation.

Adding stylistic block supports to the Spacer would further encourage less-than-ideal use of the Spacer block. If styling empty space is truly needed, there's the option of the Group block.

Given there is already a means of achieving the end goal using the Group block and a debatable benefit to adding supports to the Spacer block, I'd strongly recommend closing this PR again until a consensus is reached on the issue @hanneslsm created: https://github.com/WordPress/gutenberg/issues/69263.

If there is overwhelming support or desire for this, it will happen. 

Given this PR was only spun up due to the design tool consistency efforts rather than a genuine upswell of feature requests, it suggests to me that there isn't that overwhelming support. At least, not yet.

Once something is released we're stuck having to provide backward compatibility support for it. For that reason alone, can we hold off here until it's clear that's a price we want to pay?

--- 

### TL;DR for adopting block supports on Spacer

_This list isn't exhaustive but illustrates enough to support holding off on Spacer block supports_

**Pros**:
- A few people want more ways to style things

**Cons**:
- We are stuck having to provide backward compatibility for these supports on the block
- There are already other means of achieving the desired styles e.g. Group with min-height or padding etc.
- Leads to less semantic markup i.e. unneeded DOM elements 
- Potential for more confusion for end users
- Styling of a "negative space" block doesn't make logical sense
- The styling problem might be better handled at a design level rather than overloading the block's purpose

Let's close this PR until there is consensus on a path forward (if at all) over on https://github.com/WordPress/gutenberg/issues/69263

**Comment by t-hamano** [comment]
Ok, let's close this PR for now and continue the discussion in #69263.

@benazeer-ben, thank you for your efforts! If you have any opinions, I would be happy if you could leave a comment on #69263.

**Comment by hanneslsm** [comment]
Thanks @aaronrobertshaw for taking the time to share your take on it! I'll answer in the new issue soon. Much appreciated!
